magento/magento2
# #39218 - Custom Date Format Not Respected

### Description (*)


Locale settings in Date.php:82, Magento\Ui\Component\Form\Element\DataType\Date->prepare() is overriding incoming components $config dateFormat entry.

Logs below display how custom component dateFormat xml config gets overridden in prepare() function via Locale settings.  First entry is with en_US and 2nd log is with de_DE validating issue despite locale.

```
[2024-10-10T08:26:26.221850+00:00] main.INFO: PREPARE JS Config: {"extends":"hidden"} [] []

[2024-10-10T08:26:26.222765+00:00] main.INFO: PREPARE Component Config: {"dataScope":"date_from","component":"Magento_Ui\/js\/form\/element\/date","template":"ui\/form\/field","formElement":"date","label":"Date From","dataType":"date","validation":{"required-entry":true,"validate-date":"YYYY-MM-DD"},"notice":"Enter the starting date to test the widget","options":{"showsTime":false,"dateFormat":"yyyy-MM-dd","inputDateFormat":"yyyy-MM-dd","outputDateFormat":"yyyy-MM-dd","pickerDateTimeFormat":"yyyy-MM-dd","pickerDateFormat":"yyyy-MM-dd","storeLocale":"en_US"},"storeTimeZone":"America\/New_York"} [] []
```
===============================
```
[2024-10-11T00:33:59.318103+00:00] main.INFO: PREPARE Component Config: {"dataScope":"date_from","component":"Magento_Ui\/js\/form\/element\/date","template":"ui\/form\/field","formElement":"date","label":"Date From","dataType":"date","validation":{"required-entry":true,"validate-date":"yyyy-MM-dd"},"notice":"Enter the starting date to test the widget","options":{"showsTime":false,"dateFormat":"dd.MM.yy","inputDateFormat":"yyyy-MM-dd","outputDateFormat":"yyyy-MM-dd","pickerDateTimeFormat":"yyyy-MM-dd","pickerDateFormat":"yyyy-MM-dd","storeLocale":"de_DE"},"storeTimeZone":"Europe\/Luxembourg"} [] []
```

The dateFormat config option was not in the [original steps to reproduce XML](https://github.com/magento/magento2/issues/39218#issue-2549639762) but it is the primary option used for Validation and Frontend Display via Calendar Widget and DatePicker js files, when adding dateFormat to Ui component form xml issue still occurred.

Screenshot of issue prior to applied fix:
<img width="1157" alt="image" src="https://github.com/user-attachments/assets/a6ab2e3e-a30c-4d5e-8571-408d01d55da2">

This fix adds null coalesce check if the components config option already has a set `dateFormat` if so then do not apply system locale `dateFormat` for incoming component else use `$this->localeDate->getDateFormat();`

Additional notes:
- Developer must provide `validate-date`  format to ensure proper usage.  
- Developer should provide an option in their plugin config to allow Store User to enable an override to use System dateFormat to apply instead of plugin specific format.

### Related Pull Requests
This one

### Fixed Issues (if relevant)

1. Fixes magento/magento2#39218

### Manual testing scenarios (*)


1. Tested against [i39218 module](https://github.com/magento/magento2/issues/39218#issuecomment-2379172022)

2. Also tested against my own module, added Save component to store into database confirming entire process works as intended:
  - Installed plugin [ItsLeoFisher_custom_test_module.zip](https://github.com/user-attachments/files/17468335/ItsLeoFisher_custom_test_module.zip)
  - Visit Content > Calendar Debug > Select A Date.  
      - Observe format should be `yyyy-MM-dd` as set in view/adminhtml/ui_component/calendar_debug_form.xml instead of `mm/dd/yyyy` as default for en_US.
  - Change My Account admin locale to German (Germany) 
  - Ran: `bin/magento setup:static-content:deploy -f de_DE && bin/magento cache:clean`
  - Selected date from plugin page, date format was `yyyy-MM-dd` instead of `mm.dd.yyyy`, notice the periods in-between date groups which is default for Germany de_DE.

Output after applied fix:
```
[2024-10-11T00:41:59.077201+00:00] main.INFO: PREPARE Component Config: {"dataScope":"date_from","component":"Magento_Ui\/js\/form\/element\/date","template":"ui\/form\/field","formElement":"date","label":"Date From","dataType":"date","validation":{"required-entry":true,"validate-date":"yyyy-MM-dd"},"notice":"Enter the starting date to test the widget","options":{"showsTime":false,"dateFormat":"yyyy-MM-dd","inputDateFormat":"yyyy-MM-dd","outputDateFormat":"yyyy-MM-dd","pickerDateTimeFormat":"yyyy-MM-dd","pickerDateFormat":"yyyy-MM-dd","storeLocale":"en_US"},"storeTimeZone":"Europe\/Luxembourg"} [] []
```

Also, tested against other Locale settings via My Account > Account Setting (admin) and works as expected.

### Questions or comments

I was also noodling over if Magento Core should add another XML config option to the date component to display an Override Checkbox, so that the burden is off the developer to develop this configuration and allows flexibility to the end User to use the Magento locale dateFormat instead of the modules explicitly defined dateFormat.

Screenshot example:
<img width="1150" alt="image" src="https://github.com/user-attachments/assets/e56986b0-02cc-45eb-88a4-59738fed1b73">

Possible changes needed for new config proposal:
- Setting a new option within definition.map.xml: `<item name="useSiteLocale" type="boolean" xsi:type="xpath">settings/useSiteLocale</item>` set to false, as this can be enabled by the developer.  
- Add checkbox code to date.html with a KO binding to hide/display the checkbox
- Add observable in the date.js when checkbox is interacted with

If this is something worth while or should be Cores responsibility, I can create a ticket.   I wanted to bring this up in Slack but was not able to join.  If an invite can be sent to lcfisher15@outlook.com I can bring it up in the channel.

### Contribution checklist (*)
 - [] All new or changed code is covered with unit/integration tests (if applicable)
